src/machine/machine-dbus.c

```diff
@@ -621,23 +621,44 @@ int bus_machine_method_open_shell(sd_bus_message *message, void *userdata, sd_bu
         if (r < 0)
                 return r;
         user = empty_to_null(user);
-        if (isempty(path))
-                path = "/bin/sh";
-        if (!path_is_absolute(path))
-                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Specified path '%s' is not absolute", path);
-
-        r = sd_bus_message_read_strv(message, &args);
+        r = sd_bus_message_read_strv(message, &v);
         if (r < 0)
                 return r;
-        if (strv_isempty(args)) {
-                args = strv_free(args);
+        if (isempty(path)) {
+                char *a, *b, *get_shell;
 
-                args = strv_new(path, NULL);
+                path = "/bin/sh";
+                a = strdup("sh");
+                if (!a)
+                        return -ENOMEM;
+                b = strdup("-c");
+                if (!b)
+                        return -ENOMEM;
+                r = asprintf(&get_shell, /* This only forks one process, getent, and checks for its existance. */
+                             "shell=$(getent passwd %s 2>/dev/null | { IFS=: read _ _ _ _ _ _ x; echo \"$x\"; });\n"\
+                             "exec \"${shell:-/bin/sh}\" -", /* "-" means this is a login shell. */
+                             isempty(user) ? "root" : user);
+                if (r < 0)
+                        return -errno;
+                args = new0(char*, 3 + 1);
                 if (!args)
                         return -ENOMEM;
-
-                args[0][0] = '-'; /* Tell /bin/sh that this shall be a login shell */
+                args[0] = a;
+                args[1] = b;
+                args[2] = get_shell;
+        } else { /* If we have args we always have path, as that is the way machinectl syntax works. */
+                args = v;
+                v = NULL;
+                if (strv_isempty(args)) {
+                        args = strv_free(args);
+
+                        args = strv_new(path, NULL);
+                        if (!args)
+                                return -ENOMEM;
+                }
         }
+        if (!path_is_absolute(path))
```

Comment:
This should be moved inside the `else` above, since `path` is unconditionally assigned to be `/bin/sh` in the other branch.